crates/task-impls/src/consensus/proposal.rs

```diff
@@ -226,8 +231,323 @@ pub async fn create_and_send_proposal<TYPES: NodeType>(
 
     async_sleep(Duration::from_millis(round_start_delay)).await;
     broadcast_event(
-        Arc::new(HotShotEvent::QuorumProposalSend(message.clone(), pub_key)),
+        Arc::new(HotShotEvent::QuorumProposalSend(
+            message.clone(),
+            public_key,
+        )),
         &event_stream,
     )
     .await;
 }
+
+pub fn validate_proposal_view_and_certs<TYPES: NodeType>(
+    proposal: &Proposal<TYPES, QuorumProposal<TYPES>>,
+    sender: TYPES::SignatureKey,
+    cur_view: TYPES::Time,
+    quorum_membership: Arc<TYPES::Membership>,
+    timeout_membership: Arc<TYPES::Membership>,
+) -> Result<()> {
+    let view = proposal.data.get_view_number();
+    ensure!(
+        view > cur_view,
+        "Proposal is from an older view {:?}",
+        proposal.data.clone()
+    );
+
+    let view_leader_key = quorum_membership.get_leader(view);
+    ensure!(
+        view_leader_key == sender,
+        "Leader key does not match key in proposal"
+    );
+
+    // Verify a timeout certificate OR a view sync certificate exists and is valid.
+    if proposal.data.justify_qc.get_view_number() != view - 1 {
+        if let Some(received_proposal_cert) = proposal.data.proposal_certificate.clone() {
+            match received_proposal_cert {
+                ViewChangeEvidence::Timeout(timeout_cert) => {
+                    ensure!(timeout_cert.get_data().view == view - 1, "Timeout certificate for view {} was not for the immediately preceding view", *view);
+                    ensure!(
+                        timeout_cert.is_valid_cert(timeout_membership.as_ref()),
+                        "Timeout certificate for view {} was invalid",
+                        *view
+                    );
+                }
+                ViewChangeEvidence::ViewSync(view_sync_cert) => {
+                    ensure!(
+                        view_sync_cert.view_number == view,
+                        "View sync cert view number {:?} does not match proposal view number {:?}",
+                        view_sync_cert.view_number,
+                        view
+                    );
+
+                    // View sync certs must also be valid.
+                    ensure!(
+                        view_sync_cert.is_valid_cert(quorum_membership.as_ref()),
+                        "Invalid view sync finalize cert provided"
+                    );
+                }
+            }
+        } else {
+            bail!(
+                "Quorum proposal for view {} needed a timeout or view sync certificate, but did not have one",
+                *view);
+        };
```

Comment:
rather than `if let Some() = ... { } else { bail!() }`, I think this can just be `let received_proposal_cert = proposal.data.proposal_certificate.clone().context("...");`.